src/node/internal/internal_http_client.ts

```diff
@@ -0,0 +1,420 @@
+  #incomingMessage?: IncomingMessage;
+  #timer: number | null = null;
+
+  public _ended: boolean = false;
+
+  public timeoutCb?: VoidFunction | undefined;
+  public timeout?: number;
+  public method: string = 'GET';
+  public path: string;
+  public host: string;
+  public protocol: string = 'http:';
+  public port: string = '80';
+  public joinDuplicateHeaders: boolean | undefined;
+
+  public constructor(
+    input: string | URL | RequestOptions | null,
+    options?: RequestOptions | IncomingMessageCallback,
+    cb?: IncomingMessageCallback
+  ) {
+    super();
+
+    if (typeof input === 'string') {
+      input = urlToHttpOptions(new URL(input));
+    } else if (isURL(input)) {
+      // url.URL instance
+      input = urlToHttpOptions(input);
+    } else {
+      cb = options as IncomingMessageCallback;
+      options = input as RequestOptions;
+      input = null;
+    }
+
+    if (typeof options === 'function') {
+      cb = options;
+      options = input ?? {};
+    } else {
+      options = Object.assign(input ?? {}, options);
+    }
+
+    if (options.path) {
+      const path = String(options.path);
+      if (INVALID_PATH_REGEX.test(path)) {
+        throw new ERR_UNESCAPED_CHARACTERS('Request path');
+      }
+    }
+
+    // We don't support any other protocol than http: but
+    // node.js supports different protocols according to the default agent.
+    if (options.protocol !== undefined) {
+      if (options.protocol !== 'http:') {
+        throw new ERR_INVALID_PROTOCOL(options.protocol as string, 'http:');
+      }
+    }
+
+    const defaultPort = options.defaultPort || 80;
+
+    const port = (options.port = options.port || defaultPort || 80);
+    this.port = port.toString();
+    const host = (options.host =
+      validateHost(options.hostname, 'hostname') ||
+      validateHost(options.host, 'host') ||
+      'localhost');
+
+    const setHost =
+      options.setHost !== undefined
+        ? Boolean(options.setHost)
+        : options.setDefaultHeaders !== false;
+    if (options.timeout !== undefined)
+      this.timeout = getTimerDuration(options.timeout, 'timeout');
+
+    const signal = options.signal;
+    if (signal) {
+      // @ts-expect-error TS2379 Type inconsistency
+      addAbortSignal(signal, this);
+    }
+    let method = options.method;
+    const methodIsString = typeof method === 'string';
+    if (method != null && !methodIsString) {
+      throw new ERR_INVALID_ARG_TYPE('options.method', 'string', method);
+    }
+
+    if (methodIsString && method) {
+      if (!checkIsHttpToken(method)) {
+        throw new ERR_INVALID_HTTP_TOKEN('Method', method);
+      }
+      method = this.method = method.toUpperCase();
+    } else {
+      method = this.method = 'GET';
+    }
+
+    const maxHeaderSize = options.maxHeaderSize;
+    if (maxHeaderSize !== undefined) {
+      validateInteger(maxHeaderSize, 'maxHeaderSize', 0);
+      throw new ERR_OPTION_NOT_IMPLEMENTED('options.maxHeaderSize');
```

Comment:
`maxHeaderSize` seems like it could be okay to ignore for compatibility.

Reply:
We can always remove these errors if they're too disruptive but we can not readd them due to backward compatibility limitations.